cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -936,6 +937,101 @@ impl Inst {
             srcloc,
         }
     }
+
+    /// Choose which instruction to use for moving a value between registers.
+    pub(crate) fn r#move(
+        ty: Type,
+        from_reg: Reg,
+        to_reg: Writable<Reg>,
+        _srcloc: Option<SourceLoc>,
+    ) -> Inst {
+        if is_movable_in_scalar_reg(ty) {
+            Inst::mov_r_r(ty.bits() == 64, from_reg, to_reg)
+        } else {
+            let opcode = match ty {
+                types::F32 => SseOpcode::Movss,
+                types::F64 => SseOpcode::Movsd,
+                types::F32X4 => SseOpcode::Movaps,
+                types::F64X2 => SseOpcode::Movapd,
+                _ if ty.is_vector() && ty.bits() == 128 => SseOpcode::Movdqa,
+                _ => unimplemented!("unable to move type: {}", ty),
+            };
+            Inst::xmm_unary_rm_r(opcode, RegMem::reg(from_reg), to_reg)
+        }
+    }
+
+    /// Choose which instruction to use for loading a register value from memory. This currently
+    /// zero extends any loads smaller than 64 bits but should eventually support sign extension
+    /// (TODO).
```

Comment:
I like having helpers on `Inst` here. However, I think the extension should be an optional behavior; it's easy to imagine reaching for `Inst::load()` in the future for some general use-case and getting an extension op when it wasn't really needed (e.g. most narrow arithmetic doesn't care about high bits).

Could we define an enum for extension mode, something like the aarch64 backend's `NarrowValueMode`, to use here?

Reply:
I added an `ExtKind` enum; with that in place we should now be able to remove `xmm_mov` and simplify the lowering considerably in a future PR.